AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/MutationErrorRefinement.dfy

```diff
@@ -125,23 +125,51 @@ module {:options "/functionSyntax:4" } MutationErrorRefinement {
-    if (opaqueKmsError?.Some? && kmsErrorMessage?.Some?) {
-      var hasReEncryptFrom? := String.HasSubString(kmsErrorMessage?.value, "ReEncryptFrom");
-      var hasReEncryptTo? := String.HasSubString(kmsErrorMessage?.value, "ReEncryptTo");
-      if (hasReEncryptFrom?.Some?) {
-        return Types.MutationFromException(
-            message := "Key Management denied access based on the original properties."
-            + " Mutation is halted. Check access to KMS ARN: " + item.KmsArn + "."
-            + "\n" + errorContext
-          );
-      }
-      if (hasReEncryptTo?.Some?) {
-        return Types.MutationToException(
-            message := "Key Management denied access based on the terminal properties."
-            + " Mutation is halted. Check access to KMS ARN: " + terminalKmsArn  + "."
-            + "\n" + errorContext
-          );
+    if (kmsWithMsg && knownKmsStrat) {
+      match kmsOperation {
+        case "ReEncrypt" =>
+          var hasReEncryptFrom? := String.HasSubString(kmsErrorMessage?.value, "ReEncryptFrom");
+          var hasReEncryptTo? := String.HasSubString(kmsErrorMessage?.value, "ReEncryptTo");
+          if (hasReEncryptFrom?.Some?) {
+            return Types.MutationFromException(
+                message := "Key Management denied access based on the original properties."
+                + " Mutation is halted. Check access to KMS ARN: " + item.KmsArn + "."
+                + "\n" + errorContext
+              );
+          }
+          if (hasReEncryptTo?.Some?) {
+            return Types.MutationToException(
+                message := "Key Management denied access based on the terminal properties."
+                + " Mutation is halted. Check access to KMS ARN: " + terminalKmsArn  + "."
+                + "\n" + errorContext
+              );
+          }
+        case "Decrypt/Encrypt" =>
+          var hasDecrypt? := String.HasSubString(kmsErrorMessage?.value, "Decrypt");
+          var hasEncrypt? := String.HasSubString(kmsErrorMessage?.value, "Encrypt");
+          if (hasDecrypt?.Some?) {
+            return Types.MutationFromException(
+                message := "Key Management denied access based on the original properties."
+                + " Mutation is halted. Check decrypt access to KMS ARN: " + item.KmsArn + "."
+                + "\n" + errorContext
+              );
+          }
+          if (hasEncrypt?.Some?) {
+            return Types.MutationToException(
+                message := "Key Management denied access based on the terminal properties."
+                + " Mutation is halted. Check encrypt access to KMS ARN: " + terminalKmsArn + "."
+                + "\n" + errorContext
+              );
+          }
+        case _ =>
+          // This will never happen
+          return Types.KeyStoreAdminException(
+              message := "Unknown KMS Operation used for Mutation operation: " + kmsOperation
+              + ".\n" + errorContext
+            );
```

Comment:
This is wrong.
I should have caught this earlier.
This case WILL happen if the two booleans fire but no match occurs,
which is a possibility.
In which case, we should fall through to the final return on 176.

Reply:
I was wrong (again) when I wrote this comment.
This should be impossible, as the operation matched.